src/libsystemd-network/sd-dhcp-client.c

```diff
@@ -1561,11 +1602,17 @@ static int client_parse_message(
         lease->next_server = message->siaddr;
         lease->address = message->yiaddr;
 
+        if (client->bootp)
+                lease->lifetime = USEC_INFINITY;
+
+        if (lease->server_address == 0 && !client->bootp)
```

Comment:
@colin-foster-in-advantage this appears to be the root cause of `test-sd-dhcp-lease.c` yielding a false PASS when testing with a BOOTP reply that does not contain DHCP OPTION 53 (Server Identifier). I believe that test should fail when `server_address` is `0`, so `&& !client-bootp` is incorrect here.

An alternative would be to ensure the rest of `networkd` does not require `server_address` when `BOOTP=yes`, as in `networkd-dhcp4.c`. However, I worry that would be a bigger and higher-risk change, and is not correct. 

IMHO the way `server_address` is used in `networkd-dhcp4.c` is consistent with the BOOTP definition of `siaddr` (which is analogous to DHCP Server Identifier), not the DHCP updated definition of `siaddr` ("Next Server Address"). DHCP Server Identifier is not required by BOOTP, so an implementation that errors when it is absent would be incorrect.